File: libsrc/twopoint.f90
module twopoint

implicit none
private get1Dvel
public  two_point_td

contains

function get1Dvel(depth,nl,d,v)

implicit none
real                           :: get1Dvel
real,               intent(IN) :: depth
integer,            intent(IN):: nl
real, dimension(:), intent(IN):: d
real, dimension(:), intent(IN):: v

integer    i
real       vel
real       vair                        !velocity of sound wave in air
parameter (vair = 0.300)

if (depth .lt. d(1)) then
    get1Dvel=vair
    return
end if

vel=vair
do i=nl,1,-1
    if (depth .ge. d(i)) exit  !break cycle
end do
if (i .gt. 0) then
    vel=v(i)
else
    vel=vair
end if

get1Dvel=vel

end function get1Dvel

subroutine two_point_td(ph,srcxyz,obsxyz,td,toa)
! Compute travel time, derivations and take off angle
! z-axes downward
use const_raddeg
use layers
implicit none

real,dimension(3), intent(IN)  :: srcxyz    !coordinates of the hypocenter
real,dimension(3), intent(IN)  :: obsxyz    !coordinates of the station
character(*), intent(IN)   :: ph        !seismic phase
!      td(1) travel time
!      td(2:4) derivation on the x,y,z in the hypocenter
real,dimension(4), intent(OUT) :: td
! toa - ray take off angle of the hypocenter relative to the z-axis
real, intent(OUT)              :: toa

interface
include 'rt1Dlayers.fh'
end interface

   type(cLayers), pointer  :: lptr=>null()
   integer :: nl        !number of layers
   real(8) toas
   real(8) toac
   real(8) raypar
   integer type_of_wave
   real    v_hypo       !velocity in hypoc.
   real    delta        !distance
   real    tt           !travel time 1D

   lptr=>get_layers(ph)
   if (associated(lptr) .eqv. .false.) then
      td(1)=-1.0
      td(2:4)=[0.0,0.0,0.0]
      return
   end if
   nl=lptr%n_layers
   delta = sqrt((obsxyz(1)-srcxyz(1))**2+(obsxyz(2)-srcxyz(2))**2)
   if (delta .lt. 0.0001) then
      delta = 0.0001
   endif

!  ray traceing in 1D
   call raytr1D (delta,srcxyz(3), nl,lptr%z,lptr%v, tt, raypar,type_of_wave)

!  Get the velocity at the hypocenter to calculate the derivative dt/dz
   v_hypo = get1Dvel(srcxyz(3),nl,lptr%z,lptr%v)

!  travel time
   td(1) = tt

!  derivation by x  of travel time
   td(2)=real(-raypar*(obsxyz(1)-srcxyz(1))/delta)

!  derivation by y
   td(3)=real(-raypar*(obsxyz(2)-srcxyz(2))/delta)

!  derivation by z
   td(4)=0.0
   toas=raypar*v_hypo
   toac = dsqrt(1D0-toas*toas)
   if (isnan(toac)) then
      toas=1.D0
      toac=0.D0
   end if
   if (type_of_wave .eq. 1) then
      toac=-toac
   end if
   td(4)=real(toac/v_hypo)

!  the directional cosine of the ray to the z-axis
   toa=real(acos(-toac)*RAD2DEG)
   return
end subroutine two_point_td

end module twopoint

File: libsrc/twopoint.cpp
#include <cmath>
#include <vector>
#include <iostream>
#include <memory>
#include <map>
#include <algorithm>

#include "rt1Dlayers.h" // contains the definition of raytr1D
/*
struct TLayers{
    int n_layers;
    std::vector<float> z;
    std::vector<float> v;
};
*/
#include "layers.h" // contains the definition of get_layers



// vector std::vector<float> is sorted in ascending order
// use std::upper_bound and std::distance and find the index of element in the vector x
// corresponding to the float number 'value'
// This function returns the index of the largest element in the vector x that is less than or equal to value.
// If all elements in x are less than value, it returns the last index of the vector.
// If all elements in x are greater than value, it returns -1.
// If the vector is empty, it returns -1.
// If the vector contains only one element, it returns 0 if that element is less than or equal to value, otherwise it returns -1.
// If the vector contains two elements, it returns 0 if the first element is less than or equal to value and the second element is greater than value, otherwise it returns -1.
int find_index(const std::vector<float>& x, float value) {
    if (x.empty()) {
	return -1; // Return -1 for empty vector
    }
    if (x.size() == 1) {
	return (x[0] <= value) ? 0 : -1; // Return 0 or -1 for single element vector
    }

    auto it = std::upper_bound(x.begin(), x.end(), value);  //if (!(val<*it))
    int index = std::distance(x.begin(), it) - 1;
    return (index >= 0 && index < static_cast<int>(x.size())) ? index : -1;
}


double get1Dvel(double depth, int nl, const std::vector<float>& d, const std::vector<float>& v) {
    double vel;
    const double vair = 0.300;

    if (depth < d[0]) {
        return vair;
    }

    vel = vair;
    //int i;
    //for (i = nl - 1; i >= 0; --i) {
    //    if (depth >= d[i]) break;
    //}
    
    int i = find_index(d, depth);
    
    if (i >= 0) {
        vel = v[i];
    } else {
        vel = vair;
    }

    return vel;
}

void two_point_td(const double (&srcxyz)[3], const double (&obsxyz)[3],
	       	const std::string ph, double (&td)[4], double& toa) {
// Function to calculate the travel time and ray parameter for a 1D layered medium
// using the raytr1D function
    std::shared_ptr<TLayers> lptr = getLayersPointer(ph);
    int nl;
    double toas;
    double toac;
    double raypar;
    double v_hypo;
    double delta;
    double tt=-1.0;
    int type_of_wave=-1;

    if (!lptr) {
        td[0] = -1.0;
        td[1] = 0.0;
        td[2] = 0.0;
        td[3] = 0.0;
        return;
    }
    nl = lptr->n_layers;
    delta = std::sqrt(std::pow(obsxyz[0] - srcxyz[0], 2) + std::pow(obsxyz[1] - srcxyz[1], 2));
    if (delta < 0.0001) {
        delta = 0.0001;
    }

    raytr1D(delta, srcxyz[2], nl, lptr->z.data(), lptr->v.data(),
		    &tt, &raypar, &type_of_wave);

    v_hypo = get1Dvel(srcxyz[2], nl, lptr->z, lptr->v);

    td[0] = tt;
    td[1] = -raypar * (obsxyz[0] - srcxyz[0]) / delta;
    td[2] = -raypar * (obsxyz[1] - srcxyz[1]) / delta;
    td[3] = 0.0;
    toas = raypar * v_hypo;
    toac = std::sqrt(1.0 - toas*toas);
    if (std::isnan(toac)) {
        toas = 1.0;
        toac = 0.0;
    }
    if (type_of_wave == 1) {
        toac = -toac;
    }
    td[3] = toac / v_hypo;

    toa = std::acos(-toac) * (180.0 / M_PI); // Convert radians to degrees
}


// main function for testing find_index and get1Dvel
// one argument is the 'value' to be searched in the vector
// test data vector x and vector v is hardcoded
// print the index of the value in the vector x
// and x[index] and v[index] to the console
// return 0
/*
int main(int argc, char* argv[]) {
    std::vector<float> x = {0.0, 1.0, 2.0, 3.0, 4.0};
    std::vector<float> v = {5.0, 6.0, 7.0, 8.0, 9.0};
    float value = 2.5;
    if (argc > 1) {
	value = std::stof(argv[1]);
    }
    int index = find_index(x, value);
    if (index >= 0) {
	std::cout << "Index: " << index << ", x[index]: " << x[index] << ", v[index]: " << v[index] << std::endl;
    } else {
	std::cout << "Value not found in the vector." << std::endl;
    }

    double vel= get1Dvel(value, x.size(), x, v);

    std::cout << "Velocity at depth " << value << ": " << vel << std::endl;

    return 0;
}
*/

